pkg/generators/clientgen/clientgen.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package clientgen
+
+import (
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"k8s.io/code-generator/cmd/client-gen/types"
+	"k8s.io/gengo/namer"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-tools/pkg/genall"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+
+	"github.com/kcp-dev/code-generator/pkg/internal/clientgen"
+	"github.com/kcp-dev/code-generator/pkg/parser"
+	"github.com/kcp-dev/code-generator/pkg/util"
+)
+
+type Generator struct {
+	// Name is the name of this client-set, e.g. "kubernetes"
+	Name string `marker:""`
+
+	// HeaderFile specifies the header text (e.g. license) to prepend to generated files.
+	HeaderFile string `marker:",optional"`
+
+	// Year specifies the year to substitute for " YEAR" in the header file.
+	Year string `marker:",optional"`
+
+	// SingleClusterClientPackagePath is the root directory under which single-cluster-aware clients exist.
+	// e.g. "k8s.io/client-go/kubernetes"
+	SingleClusterClientPackagePath string `marker:""`
+
+	// OutputPackagePath is the root directory under which this tool will output files.
+	// e.g. "github.com/kcp-dev/client-go/clients"
+	OutputPackagePath string `marker:""`
+
+	// APIPackagePath is the root directory under which API types exist.
+	// e.g. "k8s.io/api"
+	APIPackagePath string `marker:"apiPackagePath"`
+}
+
+func (Generator) RegisterMarkers(into *markers.Registry) error {
+	return markers.RegisterAll(into,
+		parser.GenclientMarker,
+		parser.NonNamespacedMarker,
+		parser.GroupNameMarker,
+		parser.NoVerbsMarker,
+		parser.ReadOnlyMarker,
+		parser.SkipVerbsMarker,
+		parser.OnlyVerbsMarker,
+	)
+}
+
+// Generate will generate clients for all types that have generated clients.
+func (g Generator) Generate(ctx *genall.GenerationContext) error {
+	var headerText string
+
+	if g.HeaderFile != "" {
+		headerBytes, err := ctx.ReadFile(g.HeaderFile)
+		if err != nil {
+			return err
+		}
+		headerText = string(headerBytes)
+	}
+	headerText = strings.ReplaceAll(headerText, " YEAR", " "+g.Year)
+
+	groupVersionKinds, err := parser.CollectKinds(ctx)
+	if err != nil {
+		return err
+	}
+
+	groupInfo := toGroupVersionInfos(groupVersionKinds)
+
+	clientsetDir := filepath.Join("clients", "clientset", "versioned")
```

Comment:
We could probably drop `versioned` from the path if we wanted

Reply:
Yeah, really didn't know what the `versioned` and `typed` bits were all about - maybe more useful when k8s had internal and external clients in one spot?

Reply:
Let's revisit in the future if we find it's hurting DX - right now this is parallel with client-gen output.